packages/api/src/apps/api/app.ts

```diff
@@ -97,25 +98,46 @@ async function run(env: ProcessEnv) {
   // expose calls through express
   await Express({ port: Number(env.EXPRESS_PORT) }, actions);
 
-  // main update loop, update every block
-  provider.on("block", (blockNumber: number) => {
+  // break all state updates by block events into a cleaner function
+  async function updateByBlock(blockNumber: number) {
+    await services.blocks.handleNewBlock(blockNumber);
     // dont do update if this number or blocks hasnt passed
-    services.blocks.handleNewBlock(blockNumber).catch(console.error);
     if (blockNumber - appState.lastBlockUpdate >= updateBlocks) {
-      services.registry(appState.lastBlock, blockNumber).catch(console.error);
-      services.emps(appState.lastBlock, blockNumber).catch(console.error);
+      // update everyting
+      await services.registry(appState.lastBlock, blockNumber);
+      await services.emps(appState.lastBlock, blockNumber);
+      await services.erc20s.update();
+      await services.empStats.update();
+
       appState.lastBlockUpdate = blockNumber;
-      services.erc20s.update().catch(console.error);
-      services.empStats.update().catch(console.error);
     }
     appState.lastBlock = blockNumber;
-    services.blocks.cleanBlocks(oldestBlock).catch(console.error);
+    await services.blocks.cleanBlocks(oldestBlock);
+  }
+
+  // main update loop, update every block
+  provider.on("block", (blockNumber: number) => {
```

Comment:
Are we assuming that this is the update function? Is there a way to generalize this so it works inside a cron job? We don't need to do this now, but I think we should be mindful of architecting this in a way that traps us into running this in a long-running process.

For instance, I could imagine us wanting to have different trigger types:
1. New blocks/internal timer
2. An http post (we might separate the post calls for price updates vs blockchain updates if they need to run at diff frequencies).
3. A single-run script that executes an update and terminates.

I'm not saying we should _build_ these today. We're just building this API in a pretty generalized way, so we should be sure to add generalizations for use cases that we think are probable.

Reply:
This particular file is like the top level composition of all the lego blocks (services, tables, timers).  Each service right now has a fairly unopinionated `update` function which can be called by anything, so you could add private http calls to update, add a new app which just calls update and exits, or change the interval on which they are called.  Once we have a shared data store, we can start to get more sophisticated with the design.

Reply:
Yeah, totally agree on your point about the services being unopinionated.

I guess my main point is that this file seems to couple some pretty different pieces of functionality: the way the update is triggered, the way those updates are stored, and the way that the store is queried.